### ISSUE TITLE
Presets panel and rendering animation

### ISSUE THREAD

**Issue by Durman** [nortikin/sverchok#4316]
## Problem statement

There is strange problem which appears only during rendering animation when the Presets panel is opened. The problem is that any data can't be written into attributes of Blender ID objects. The problem does not happen in each frame.

## Steps to reproduce

For last stable release it's enough to have tree with Frame info node to reproduce the problem.
```
2021-08-30 13:21:23,431 [WARNING] sverchok.core.sockets 568: Socket='Current Frame' of node='Frame info' can't update number of objects on the label. Cause is 'Writing to ID classes in this context is not allowed: NodeTree, NodeTree datablock, error setting SvStringsSocket.objects_number'
Failed to update: Writing to ID classes in this context is not allowed: NodeTree, NodeTree datablock, error setting SverchCustomTreeType.error nodes
```

In new update system the problem does not appear if update made via timer but if update is made within frame change handler similar error is rised.

## Sverchok version

This problem is in both last stable release and v1.0

**Comment by zeffii** [comment]
@Durman  this also gives that warning? 
https://gist.github.com/3e79200c56819938ee1c7382db40879f  i can't get that error to appear.

**Comment by Durman** [comment]
Yes sometimes

https://user-images.githubusercontent.com/28003269/131339055-1bcedfb2-54ac-42e9-8052-cf7605707e00.mp4

**Comment by zeffii** [comment]
doesn't do that on current master here. very unsatisying.

i imagine if we stopped updating `socket.objects_number` while Blender is in a render-state then there will simply be other properties that will cause the same error.

**Comment by zeffii** [comment]
i see what you mean now.  the `presets panel` must be open, to get this error.

**Comment by zeffii** [comment]
- we could add a loop through the nodeviews, and foldup any sverchok-presets-panels in a `render_init` handler, 
- https://docs.blender.org/api/current/bpy.app.handlers.html#note-on-altering-data

**Comment by Durman** [comment]
Hidding the panel seems to me is not very good solution. I think UI code of the panel uses something that cause the problem. Would be nice to find exact string commenting of which would stop the problem.

**Comment by zeffii** [comment]
if no better solution is found. hiding (or folding up) the panel is an option. Realistically we aren't going to be modifying the nodetrees anyway while rendering. If people want to continue using Blender while rendering, they may better start a new instance of Blender. There's a good reason why Blender has a "lock interface" feature while rendering. 

anyway, this is "not a hill worth dying on."

we must be doing something inside the draw function that directly or indirectly causes ID classes to be updated.

**Comment by zeffii** [comment]
i don't think we can close the panel ( `space.show_region_toolbar = False`) from a `render_init` handler, even if we wanted to. Not like this anyway.
```python
import bpy

def render_init_func(scene):
    for window in bpy.context.window_manager.windows:
        screen = window.screen
        for area in screen.areas:
            if area.type == 'NODE_EDITOR':
                for space in area.spaces:
                    ng = space.edit_tree if hasattr(space, "edit_tree") else None
                    if not ng:
                        continue
                    if ng.bl_idname in {"SverchCustomTreeType"} and space.type == 'NODE_EDITOR':
                        try:
                            space.show_region_toolbar = False
                            continue
                        except Exception as err:
                            print("show_region_toolbar:", err)

if __name__ == '__main__':
    bpy.app.handlers.render_init.append(render_init_func)
```

this will close Blender entirely. But the code does work if triggered simply via
```python
render_init_func(bpy.context.scene)
```

**Comment by zeffii** [comment]
https://github.com/nortikin/sverchok/pull/4319 ?

**Comment by zeffii** [comment]
if the draw function returns before this part there are no errors/warning
```python
        return
        if not panel_props.search_text:
            layout.prop(panel_props, 'category', text='')
```
but if you do return directly after it
```python
        if not panel_props.search_text:
            layout.prop(panel_props, 'category', text='')
        return
```
then the errors/warnings are as usual.

**Comment by zeffii** [comment]
it's tempting to rewrite `panel_props.search_text` to use an update function.  I think .. i hope.. that the current method uses a sideeffect of the fact that the draw function is called a lot when the user moves a mouse around in the panel (and when typing). If the `search_text` property only updates when the user changes it, then I think this NodeTree class ID isn't updated unless user types something.

only one way to find out.

**Comment by Durman** [comment]
According your investigation my guess is that the problem is in the enum property.

https://github.com/nortikin/sverchok/blob/5a098e47b62614ee95babd199c319e18f1d302bc/ui/presets.py#L395-L398

The thing is that items of enum properties should be saved somewhere in global space of a module. It similar to this issue #4237. It worth trying to rewrite `get_category_items_nonempty` function.

**Comment by zeffii** [comment]
it certainly looks that way.

**Comment by zeffii** [comment]
One way to make caching possible is to add a "refesh" operator. 
- if user makes changes to file system, refresh will only need to iterate the filesystem once and dump to the cache
- any other operator / property that adds/removes categories (or removes items from categories.. like `deletepreset`) would trigger the `refresh` behind the scenes.

@portnov any opinions?  UserPreferes->Add-ons  must also manually refresh to detect drag/dropped files.

**Comment by portnov** [comment]
IMHO we can safely assume no one wants to create / remove a directory under presets/ and want to see the new menu item in blender momentarily. No problem in restarting blender or pressing F8 in such a case. So, we can populate cache at register() and only update it when manually creating / removing a category from blender UI. Just need to be very accurate to do not miss anything.

**Comment by zeffii** [comment]
> No problem in restarting blender or pressing F8 in such a case.

a small refresh button could be a nice UI touch. It would be minimal work compared to the rest of the needed changes. Else it could be a side effect of the _manage presets_ boolean state change. F8 seems overkill. but let's see.

**Comment by Durman** [comment]
Example of possible solution

```py
def cache_enum(func):
    cache = [None, None]

    def inner():
        del cache[0]
        enum = func()
        cache.append(enum)
        return enum

    return inner


@cache_enum  # can be used for other functions too
def generate_list_example():
    l = []
    for i in range(10):
        l.append(i, 'name' + str(i), 'descr' + str(i), 0, i)
    return l
```
I did not test this.

**Comment by Durman** [comment]
I was able to reproduce the error even without the enum to be displayed. The problem might be in something elese.
